### PR TITLE
[ws-manager] Replace PLIS with Kubernetes finalizer

### PR DESCRIPTION
**Pull request by csweichel**
This PR replaces the pod lifecycle independent state with a Kubernetes finalizer. 
Fixes https://github.com/gitpod-io/gitpod/issues/3335

### How to test
- Start a workspace and stop it again
- Start workspaces that fail to start (e.g. remove registry-facade)
- Inspect the unit tests

### CODE REVIEW

**Comment by geropl** [comment]
Seen the PR but won't be able to review before Monday. :+1:

**Comment by aledbf** [comment]
@csweichel why not using the approach suggested by controller-runtime with the finalizers? [example](https://github.com/gitpod-io/gitpod/pull/3331/files#diff-0a3211e8414a139738a48c2bbfa361e21506659c2bae3ba4f643dedfefc81156R41-R70)
With this change, what's the role of the `ConfigmapReconciler`? (can be removed?)

**Comment by csweichel** [comment]
> ConfigmapReconciler? can be removed?

💯  fixed now

> why not using the approach suggested by controller-runtime with the finalizers

Indeed we could go that way. I reckon there are three paths forward:
1. partially migrate to `PodReconciler`, maybe just `actOnPodEvent` or even maybe even just the bits that pertain to the finalizer
    - :+1: probably fewest LoC changed (LoC as awful proxy for "invasiveness")
    - :-1: pretty unclear behaviour, because some bits of the "feedback loop" remain in the old structure, while others move to the new. Opens the door for races.
2. completely migrate to `PodReconciler` and remove the event worker pool
    - :+1: clearest step moving towards a more "Kubernetes like" structure
    - :+1: at first glance feels like it's not too hard to do, primarily removing the event worker pool
    - :-1: monitor was written with the concurrency guarantees of the event worker pool in mind. If we remove it, we have to make sure we're not blocking the reconciliation loop
    - :-1: the event worker pool provides high levels of concurrency. To my understanding we'd be giving this up and putting everything in a single loop (also see point above), possibly degrading performance
3. feed the event worker pool using the `PodReconciler` and maintain the old `handlePodEvent`/`actOnPodEvent` structure
    - :+1: introduces the reconciler concept (but doesn't quite use it yet)
    - 👍 keeps current monitor behaviour in place because it doesn't remove the expectations towards concurrency
    - 👎 abuses the reconciler to drive an "out-of-the-ordinary" structure

I've currently implemented option 3 because it's the minimal change.
Option 2 looks like it's much closer the rest of the Kubernetes world, but may be a change for a separate PR. As part of that change, I'd very much like to see what performance impact that change has under load and ensure that monitor does not assume it can do long-running operations synchronously (afaik it doesn't).

**Review by aledbf — APPROVED** [review]

**Comment by geropl** [comment]
Starting to review...

File: components/ws-manager/pkg/manager/annotations.go [hunk 1]
```diff
@@ -145,126 +148,59 @@ func (a *annotation) Apply(dst map[string]string) (needsUpdate bool) {
...
-
-			needsUpdate = patch(plis)
-			tracing.LogEvent(span, "patch done")
-
-			err = marshalPodLifecycleIndependentState(&plisCfg, plis)
-			if err != nil {
-				return xerrors.Errorf("patch lifecycle independent state: %w", err)
+		} else {
+			n := 0
+			for _, x := range pod.Finalizers {
```
Comment (geropl) [inline components/ws-manager/pkg/manager/annotations.go:189]: This loops feels a tad to nifty.
 - Why the explicit (re-)assignment of finalizers (in the loop and later: `pod.Finalizers = pod.Finalizers[:n]`)? If we skipped that the code would become easier
Is there a Kubernetes guarantee that there is ever only one entry with a given name?

File: components/ws-manager/pkg/manager/monitor.go [hunk 2]
```diff
@@ -377,42 +372,22 @@ func (m *Monitor) actOnPodEvent(ctx context.Context, status *api.WorkspaceStatus
...
+				break
 			}
-
-			return
-		}, annotations...)
-		if err != nil {
-			return xerrors.Errorf("cannot update pod lifecycle independent state: %w", err)
+		}
+		if terminated {
+			go m.finalizeWorkspaceContent(ctx, wso)
```
Comment (geropl) [inline components/ws-manager/pkg/manager/monitor.go:390]: Do I understand correctly that we now wait for the root process to terminate before we start the backup process? What are the implications on stopping performance?

File: components/ws-manager/pkg/manager/monitor.go [hunk 3]
```diff
@@ -1326,50 +1099,6 @@ func (m *Monitor) markTimedoutWorkspaces(ctx context.Context) (err error) {
 		}
 	}
 
-	// timeout PLIS only workspaces
```
Comment (geropl) [inline components/ws-manager/pkg/manager/monitor.go:1329]: Wow. This is what relieve feels like :-D

**Review by geropl — APPROVED** [review]
code changes look good, comments are more mostly regarding understanding

build fails because of missing license headers.

**Comment by aledbf** [comment]
**Note:** this change reduces the number of events by half (configmaps). This should have an impact on API metrics.

File: components/ws-manager/pkg/manager/annotations.go (hunk 1 above)
Comment (csweichel) [reply components/ws-manager/pkg/manager/annotations.go:189]: I simply applied the filter slice trick: https://github.com/golang/go/wiki/SliceTricks#filter-in-place :)

File: components/ws-manager/pkg/manager/monitor.go (hunk 2 above)
Comment (csweichel) [reply components/ws-manager/pkg/manager/monitor.go:390]: It's been like this before as well - which is what made the PLIS necessary in the first place. Prior to finalizers we didn't have a way to keep the pod arround once the container was stopped, hence introduced the PLIS.

File: components/ws-manager/pkg/manager/annotations.go (hunk 1 above)
Comment (geropl) [reply components/ws-manager/pkg/manager/annotations.go:189]: Yes, I think I got the effect. Would have expected sth like `update := len(pod.Finalizers) != len(filteredFinalizers)` for better readability. But your mileage may vars :)

**Pull request merged**